File: src/cmdbox/services/import_service.py
import json
from dataclasses import dataclass, field
from pathlib import Path

from cmdbox.common.graph_utils import find_cycle
from cmdbox.resolve.reference_parsing import extract_references
from cmdbox.resolve.type_defs import RefKind
from cmdbox.services.command_services import CommandServices
from cmdbox.services.errors import ImportCycleError, ImportFileError
from cmdbox.services.tag_services import TagServices
from cmdbox.services.variable_services import VariableServices

_SUPPORTED_VERSIONS = {"1"}


KIND_PREFIX = {
    RefKind.COMMAND: "cmd",
    RefKind.VARIABLE: "var",
}


@dataclass
class ImportResult:
    commands_created: list[str] = field(default_factory=list)
    commands_skipped: list[str] = field(default_factory=list)
    commands_overwritten: list[str] = field(default_factory=list)
    variables_created: list[str] = field(default_factory=list)
    variables_skipped: list[str] = field(default_factory=list)
    variables_overwritten: list[str] = field(default_factory=list)
    preview: bool = False


def _label(kind: RefKind, key: str) -> str:
    return f"{KIND_PREFIX[kind]}:{key}"


def build_dependency_graph(import_data: dict) -> dict[str, list[str]]:
    """
    Builds a dependency graph from the provided import data.

    This function processes the "commands" and "variables" in the provided
    import data to construct a mapping of labels to their dependencies.
    Each label represents a command or variable, and its dependencies are
    derived from references extracted from their respective templates or values.

    Args:
        import_data (dict): A dictionary containing "commands" and
            "variables" data. Each command or variable includes an alias,
            and its references are determined by processing its template or
            value.

    Returns:
        dict[str, list[str]]: A dictionary where the keys are labels
        representing commands or variables, and the values are lists of labels
        representing their dependencies.
    """
    deps: dict[str, list[str]] = {}

    for cmd in import_data.get("commands", []):
        label = _label(RefKind.COMMAND, cmd["alias"])
        deps[label] = [
            _label(kind, key) for kind, key in extract_references(cmd["template"])
        ]

    for var in import_data.get("variables", []):
        label = _label(RefKind.VARIABLE, var["name"])
        deps[label] = [
            _label(kind, key) for kind, key in extract_references(var["value"])
        ]

    return deps


def validate_no_cycles(deps: dict[str, list[str]]) -> None:
    """
    Validates that there are no circular dependencies in the provided dependency graph.

    This function takes a dictionary where the keys represent labels and the values are lists
    of dependencies for each label. It checks if there are any cycles in the graph formed
    by the dependencies and raises an error if a cycle is detected.

    Args:
        deps (dict[str, list[str]]): A dictionary representing the dependency graph. Keys are
            the labels, and values are lists of labels each key depends on.

    Raises:
        ImportCycleError: If a circular dependency is detected in the graph.
    """

    def get_deps(label: str) -> list[str]:
        return deps.get(label, [])

    for label in deps:
        cycle = find_cycle(label, get_deps)
        if cycle is not None:
            raise ImportCycleError(cycle)


def _parse_import_file(path: str | Path) -> dict:
    try:
        text = Path(path).read_text(encoding="utf-8")
        data = json.loads(text)
    except (OSError, json.JSONDecodeError, UnicodeDecodeError) as e:
        raise ImportFileError(f"Could not read import file: {e}") from e

    if not isinstance(data, dict):
        raise ImportFileError("Import file must be a JSON object")

    version = data.get("version")
    if version not in _SUPPORTED_VERSIONS:
        supported = ", ".join(_SUPPORTED_VERSIONS)
        raise ImportFileError(
            f"Unsupported import file version: {version}. Supported versions: {supported}"
        )

    return data


class ImportService:

    def __init__(
        self,
        cmd_service: CommandServices,
        var_service: VariableServices,
        tag_service: TagServices,
    ):
        self._cmd_service = cmd_service
        self._var_service = var_service
        self.tag_service = tag_service
        self.result = ImportResult()

    def import_file(
        self, path: str | Path, overwrite: bool = False, preview: bool = False
    ) -> ImportResult:
        self.result.preview = preview
        data = _parse_import_file(path)

        deps = build_dependency_graph(data)
        validate_no_cycles(deps)

        cmd_items = data.get("commands", [])
        var_items = data.get("variables", [])

        cmd_actions: list[tuple[dict, str]] = []
        for cmd_data in cmd_items:
            alias = cmd_data["alias"]
            existing = self._cmd_service.get_command_or_none(alias)
            if existing is None:
                cmd_actions.append((cmd_data, "create"))
            elif overwrite:
                cmd_actions.append((cmd_data, "overwrite"))
            else:
                cmd_actions.append((cmd_data, "skip"))

        var_actions: list[tuple[dict, str]] = []
        for var_data in var_items:
            name = var_data["name"]
            existing = self._var_service.get_variable_or_none(name)
            if existing is None:
                var_actions.append((var_data, "create"))
            elif overwrite:
                var_actions.append((var_data, "overwrite"))
            else:
                var_actions.append((var_data, "skip"))

        self.classify_commands(cmd_actions)
        self.classify_variables(var_actions)

        if preview:
            return self.result

        tag_names = self._collect_tag_names(cmd_actions, var_actions)
        self._ensure_tag_exists(tag_names)

        self.handle_commands(cmd_actions)
        self.handle_variables(var_actions)

        return self.result

    def classify_commands(self, cmd_actions: list[tuple[dict, str]]):
        """
        Classifies commands based on their specified actions and updates the result object
        with the corresponding command names categorized into created, overwritten, or skipped.

        This is handled as a separate method with a separate iteration so
        that `preview` can still return a meaningful result.

        Args:
            cmd_actions (list[tuple[dict, str]]): A list of tuples, where each tuple
            contains a dictionary with command data and a string representing the
            action to be performed. The dictionary must include an 'alias' key.

        """
        for data, action in cmd_actions:
            alias = data["alias"]
            if action == "create":
                self.result.commands_created.append(alias)
            elif action == "overwrite":
                self.result.commands_overwritten.append(alias)
            elif action == "skip":
                self.result.commands_skipped.append(alias)

    def handle_commands(self, cmd_actions: list[tuple[dict, str]]):
        for cmd_data, action in cmd_actions:
            alias = cmd_data["alias"]

            if action == "skip":
                self.result.commands_skipped.append(alias)
                continue

            tags = cmd_data.get("tags", [])

            if action == "create":
                self._cmd_service.create_command(
                    alias=alias,
                    template=cmd_data["template"],
                    description=cmd_data.get("description", None),
                    tags=tags,
                    cwd=cmd_data.get("cwd", None),
                    shell=cmd_data.get("shell", None),
                    env=cmd_data.get("env", None),
                    timeout=cmd_data.get("timeout", None),
                )
                self.result.commands_created.append(alias)
            else:
                self.overwrite_existing_command(alias, cmd_data, tags)
                self.result.commands_overwritten.append(alias)

    def overwrite_existing_command(self, alias: str, cmd_data: dict, tags: list[str]):
        existing = self._cmd_service.get_command(alias)
        current_tags = {ct.tag.name for ct in existing.tags}
        new_tag = set(tags)
        self._cmd_service.update_command(
            alias=alias,
            template=cmd_data["template"],
            description=cmd_data.get("description", None),
            cwd=cmd_data.get("cwd", None),
            shell=cmd_data.get("shell", None),
            env=cmd_data.get("env", None),
            timeout=cmd_data.get("timeout", None),
        )
        tags_to_remove = list(current_tags - new_tag)
        tags_to_add = list(new_tag - current_tags)
        if tags_to_remove:
            self._cmd_service.remove_tags(alias, tags_to_remove)
        if tags_to_add:
            self._cmd_service.add_tags(alias, tags_to_add)

    def classify_variables(self, var_actions: list[tuple[dict, str]]):
        """
        Classifies variables based on the specified actions and updates the result object with
        the corresponding variable names categorized into created, overwritten, or skipped.

        This is handled as a separate method with a separate iteration so
        that `preview` can still return a meaningful result.

        Args:
            var_actions (list[tuple[dict, str]]): A list of tuples, where each tuple contains a
                dictionary with variable data and a string representing the associated action
                to be performed. The dictionary must include a 'name' key.
        """
        for data, action in var_actions:
            name = data["name"]
            if action == "create":
                self.result.variables_created.append(name)
            elif action == "overwrite":
                self.result.variables_overwritten.append(name)
            elif action == "skip":
                self.result.variables_skipped.append(name)

    def handle_variables(self, var_actions: list[tuple[dict, str]]):
        for var_data, action in var_actions:
            name = var_data["name"]

            if action == "skip":
                self.result.variables_skipped.append(name)
                continue

            tags = var_data.get("tags", [])

            if action == "create":
                self._var_service.create_variable(
                    name, value=var_data["value"], tags=tags
                )
                self.result.variables_created.append(name)
            else:
                self.overwrite_existing_variable(name, var_data, tags)
                self.result.variables_overwritten.append(name)

    def overwrite_existing_variable(self, name: str, var_data: dict, tags: list[str]):
        existing = self._var_service.get_variable(name)
        current_tags = {ct.tag.name for ct in existing.tags}
        new_tags = set(tags)
        self._var_service.update_variable(name, value=var_data["value"])
        tags_to_remove = list(current_tags - new_tags)
        tags_to_add = list(new_tags - current_tags)
        if tags_to_remove:
            self._var_service.remove_tags(name, tags_to_remove)
        if tags_to_add:
            self._var_service.add_tags(name, tags_to_add)

    def _collect_tag_names(
        self, cmd_actions: list[tuple[dict, str]], var_actions: list[tuple[dict, str]]
    ) -> set[str]:
        names: set[str] = set()
        for data, action in cmd_actions + var_actions:
            if action != "skip":
                names.update(data.get("tags", []))
        return names

    def _ensure_tag_exists(self, tag_names: set[str]) -> None:
        for name in tag_names:
            if not self.tag_service.tag_exists(name):
                self.tag_service.create_tag(name)

File: internal/services/import_service.go
package services

import (
	"encoding/json"
	"fmt"
	"os"
	"slices"

	"github.com/PhantomLambSoft/CmdBox/internal/repository"
	"github.com/PhantomLambSoft/CmdBox/internal/resolve"
)

var supportedVersions = []string{"1"}

type ImportResult struct {
	CommandsCreated      []string
	CommandsSkipped      []string
	CommandsOverwritten  []string
	VariablesCreated     []string
	VariablesSkipped     []string
	VariablesOverwritten []string
	Profile              string
	Preview              bool
}

// makeLabel combines a reference kind and key into a single string, separated by a colon.
func makeLabel(kind resolve.RefKind, key string) string {
	return string(kind) + ":" + key
}

func buildDependencyGraph(importData TransferDocument) map[string][]string {
	deps := make(map[string][]string)

	for _, command := range importData.Commands {
		label := makeLabel(resolve.RefKindCommand, command.Alias)
		references := resolve.ExtractReferences(command.Template)
		refLabels := make([]string, len(references))
		for i, ref := range references {
			refLabels[i] = makeLabel(ref.Kind, ref.Key)
		}
		deps[label] = refLabels
	}

	for _, variable := range importData.Variables {
		label := makeLabel(resolve.RefKindVariable, variable.Name)
		references := resolve.ExtractReferences(variable.Value)
		refLabels := make([]string, len(references))
		for i, ref := range references {
			refLabels[i] = makeLabel(ref.Kind, ref.Key)
		}
		deps[label] = refLabels
	}

	return deps
}

func findCycle(start string, getDependencies func(string) []string) []string {
	stack := make([]string, 0)
	onStack := make(map[string]struct{})
	visited := make(map[string]struct{})

	var visit func(node string) []string
	visit = func(node string) []string {
		if _, ok := onStack[node]; ok {
			cycleStart := slices.Index(stack, node)
			cycle := make([]string, len(stack[cycleStart:]))
			copy(cycle, stack[cycleStart:])
			return append(cycle, node)
		}
		if _, ok := visited[node]; ok {
			return nil
		}

		visited[node] = struct{}{}
		stack = append(stack, node)
		onStack[node] = struct{}{}

		for _, dep := range getDependencies(node) {
			if result := visit(dep); result != nil {
				return result
			}
		}

		stack = stack[:len(stack)-1]
		delete(onStack, node)
		return nil
	}

	return visit(start)
}

func validateNoCycles(deps map[string][]string) error {
	getDeps := func(label string) []string {
		return deps[label]
	}

	for label := range deps {
		cycle := findCycle(label, getDeps)
		if cycle != nil {
			return ErrImportCycle
		}
	}
	return nil
}

func parseImportFile(path string) (TransferDocument, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return TransferDocument{}, &ImportFileError{Path: path, Err: err}
	}

	var doc TransferDocument
	if err := json.Unmarshal(data, &doc); err != nil {
		return TransferDocument{}, &ImportFileError{Path: path, Err: err}
	}

	if !slices.Contains(supportedVersions, doc.Version) {
		return TransferDocument{}, &ImportFileError{Path: path, Err: ErrUnsupportedVersion}
	}

	return doc, nil
}

type ImportService struct {
	CmdService  CommandService
	VarService  VariableService
	TagService  TagService
	ProfileRepo repository.ProfileRepository
	Result      ImportResult
}

func NewImportService(
	cmdService CommandService,
	varService VariableService,
	tagService TagService,
	profileRepo repository.ProfileRepository,
) *ImportService {
	return &ImportService{
		CmdService:  cmdService,
		VarService:  varService,
		TagService:  tagService,
		ProfileRepo: profileRepo,
		Result:      ImportResult{},
	}
}

// resolveProfileName determines the effective profile name by using the provided value or fetching the active profile.
func (s *ImportService) resolveProfileName(profileName *string) (string, error) {
	if profileName != nil {
		return *profileName, nil
	}
	profile, err := s.ProfileRepo.GetActiveCommandProfile()
	if err != nil {
		return "", fmt.Errorf("getting active command profile %w", err)
	}
	return profile.Name, nil
}

type actionType string

const (
	actionCreate    actionType = "create"
	actionOverwrite actionType = "overwrite"
	actionSkip      actionType = "skip"
)

type commandAction struct {
	Data   TransferCommand
	Action actionType
}

type variableAction struct {
	Data   TransferVariable
	Action actionType
}

func (s *ImportService) ImportFile(path string, overwrite bool, preview bool, profileName *string) (ImportResult, error) {
	resolvedProfileName, err := s.resolveProfileName(profileName)
	if err != nil {
		return ImportResult{}, fmt.Errorf("resolving profile %w", err)
	}
	s.Result.Preview = preview
	s.Result.Profile = resolvedProfileName

	doc, err := parseImportFile(path)
	if err != nil {
		return ImportResult{}, fmt.Errorf("parsing import file %w", err)
	}

	deps := buildDependencyGraph(doc)
	if err = validateNoCycles(deps); err != nil {
		return ImportResult{}, err // Raw correct error, return directly
	}

	var cmdActions []commandAction
	for _, cmdData := range doc.Commands {
		existing, err := s.CmdService.GetCommandOrNil(cmdData.Alias, &resolvedProfileName)
		if err != nil {
			return ImportResult{}, fmt.Errorf("getting command %w", err)
		}

		switch {
		case existing == nil:
			cmdActions = append(cmdActions, commandAction{Data: cmdData, Action: actionCreate})
		case overwrite:
			cmdActions = append(cmdActions, commandAction{Data: cmdData, Action: actionOverwrite})
		default:
			cmdActions = append(cmdActions, commandAction{Data: cmdData, Action: actionSkip})
		}
	}

	var varActions []variableAction
	for _, varData := range doc.Variables {
		existing, err := s.VarService.GetVariableOrNil(varData.Name, &resolvedProfileName)
		if err != nil {
			return ImportResult{}, fmt.Errorf("getting variable %w", err)
		}

		switch {
		case existing == nil:
			varActions = append(varActions, variableAction{Data: varData, Action: actionCreate})
		case overwrite:
			varActions = append(varActions, variableAction{Data: varData, Action: actionOverwrite})
		default:
			varActions = append(varActions, variableAction{Data: varData, Action: actionSkip})
		}
	}

	s.classifyCommands(cmdActions)
	s.classifyVariables(varActions)

	if preview {
		return s.Result, nil
	}

	tagNames := s.collectTagNames(cmdActions, varActions)
	if err = s.ensureTagsExist(tagNames); err != nil {
		return ImportResult{}, err
	}

	if err = s.handleCommands(cmdActions, resolvedProfileName); err != nil {
		return ImportResult{}, err
	}
	if err = s.handleVariables(varActions, resolvedProfileName); err != nil {
		return ImportResult{}, err
	}

	return s.Result, nil
}

func (s *ImportService) classifyCommands(cmdActions []commandAction) {
	for _, action := range cmdActions {
		alias := action.Data.Alias
		if action.Action == actionCreate {
			s.Result.CommandsCreated = append(s.Result.CommandsCreated, alias)
		} else if action.Action == actionOverwrite {
			s.Result.CommandsOverwritten = append(s.Result.CommandsOverwritten, alias)
		} else {
			s.Result.CommandsSkipped = append(s.Result.CommandsSkipped, alias)
		}
	}
}

func (s *ImportService) handleCommands(cmdActions []commandAction, profileName string) error {
	for _, action := range cmdActions {
		if action.Action == actionSkip {
			continue
		}
		alias := action.Data.Alias
		tags := action.Data.Tags

		if action.Action == actionCreate {
			input := CreateCommandConfig{
				Alias:       alias,
				Template:    action.Data.Template,
				Description: action.Data.Description,
				Tags:        tags,
				Cwd:         action.Data.Cwd,
				Shell:       action.Data.Shell,
				Env:         action.Data.Env,
				Timeout:     action.Data.Timeout,
				ProfileName: &profileName,
			}
			if _, err := s.CmdService.CreateCommand(input); err != nil {
				return fmt.Errorf("creating command %w", err)
			}
		} else {
			if err := s.overwriteExistingCommand(action, profileName); err != nil {
				return fmt.Errorf("overwriting existing command %w", err)
			}
		}
	}
	return nil
}

func (s *ImportService) overwriteExistingCommand(action commandAction, profileName string) error {
	existing, err := s.CmdService.GetCommandWithTags(action.Data.Alias, &profileName)
	if err != nil {
		return fmt.Errorf("getting existing command %w", err)
	}
	currentTags := make(map[string]struct{}, len(existing.Tags))
	for _, tag := range existing.Tags {
		currentTags[tag.Name] = struct{}{}
	}
	newTags := make(map[string]struct{}, len(action.Data.Tags))
	for _, tag := range action.Data.Tags {
		newTags[tag] = struct{}{}
	}

	updateConfig := UpdateCommandConfig{
		Template:     &action.Data.Template,
		Description:  action.Data.Description,
		Cwd:          action.Data.Cwd,
		ClearCwd:     false,
		Shell:        action.Data.Shell,
		ClearShell:   false,
		Env:          action.Data.Env,
		ClearEnv:     false,
		Timeout:      action.Data.Timeout,
		ClearTimeout: false,
	}
	if _, err = s.CmdService.UpdateCommand(existing.Alias, &profileName, updateConfig); err != nil {
		return fmt.Errorf("updating command %w", err)
	}

	var toRemove []string
	for t := range currentTags {
		if _, found := newTags[t]; !found {
			toRemove = append(toRemove, t)
		}
	}

	var toAdd []string
	for t, _ := range newTags {
		if _, found := currentTags[t]; !found {
			toAdd = append(toAdd, t)
		}
	}

	if _, err := s.CmdService.RemoveTags(action.Data.Alias, toRemove, &profileName); err != nil {
		return fmt.Errorf("removing tags %w", err)
	}

	if _, err := s.CmdService.AddTags(action.Data.Alias, toAdd, &profileName); err != nil {
		return fmt.Errorf("adding tags %w", err)
	}

	return nil
}

func (s *ImportService) classifyVariables(varActions []variableAction) {
	for _, action := range varActions {
		name := action.Data.Name
		if action.Action == actionCreate {
			s.Result.VariablesCreated = append(s.Result.VariablesCreated, name)
		} else if action.Action == actionOverwrite {
			s.Result.VariablesOverwritten = append(s.Result.VariablesOverwritten, name)
		} else {
			s.Result.VariablesSkipped = append(s.Result.VariablesSkipped, name)
		}
	}
}

func (s *ImportService) handleVariables(varActions []variableAction, profileName string) error {
	for _, action := range varActions {
		if action.Action == actionSkip {
			continue
		}
		name := action.Data.Name
		tags := action.Data.Tags

		if action.Action == actionCreate {
			input := CreateVariableConfig{
				Name:        name,
				Value:       action.Data.Value,
				Tags:        tags,
				ProfileName: &profileName,
			}
			if _, err := s.VarService.CreateVariable(input); err != nil {
				return fmt.Errorf("creating variable %w", err)
			}
		} else {
			if err := s.overwriteExistingVariable(action, profileName); err != nil {
				return fmt.Errorf("overwriting variable %w", err)
			}
		}
	}
	return nil
}

func (s *ImportService) overwriteExistingVariable(action variableAction, profileName string) error {
	existing, err := s.VarService.GetVariableWithTags(action.Data.Name, &profileName)
	if err != nil {
		return fmt.Errorf("getting existing variable %w", err)
	}
	currentTags := make(map[string]struct{}, len(existing.Tags))
	for _, tag := range existing.Tags {
		currentTags[tag.Name] = struct{}{}
	}
	newTags := make(map[string]struct{})
	for _, tag := range action.Data.Tags {
		newTags[tag] = struct{}{}
	}

	updateConfig := UpdateVariableConfig{
		Value: &action.Data.Value,
	}
	if _, err := s.VarService.UpdateVariable(existing.Name, &profileName, updateConfig); err != nil {
		return fmt.Errorf("updating variable %w", err)
	}

	var toRemove []string
	for t, _ := range currentTags {
		if _, found := newTags[t]; !found {
			toRemove = append(toRemove, t)
		}
	}

	var toAdd []string
	for t := range newTags {
		if _, found := currentTags[t]; !found {
			toAdd = append(toAdd, t)
		}
	}

	if _, err := s.VarService.RemoveTags(existing.Name, toRemove, &profileName); err != nil {
		return fmt.Errorf("removing tags from variable %w", err)
	}

	if _, err := s.VarService.AddTags(existing.Name, toAdd, &profileName); err != nil {
		return fmt.Errorf("adding tags to variable %w", err)
	}

	return nil
}

func (s *ImportService) collectTagNames(cmdActions []commandAction, varActions []variableAction) []string {
	seen := make(map[string]struct{})
	for _, action := range cmdActions {
		if action.Action != actionSkip {
			for _, t := range action.Data.Tags {
				seen[t] = struct{}{}
			}
		}
	}
	for _, action := range varActions {
		if action.Action != actionSkip {
			for _, t := range action.Data.Tags {
				seen[t] = struct{}{}
			}
		}
	}
	names := make([]string, 0, len(seen))
	for t := range seen {
		names = append(names, t)
	}
	return names
}

func (s *ImportService) ensureTagsExist(tagNames []string) error {
	for _, tagName := range tagNames {
		tag, err := s.TagService.GetTagOrNil(tagName)
		if err != nil {
			return fmt.Errorf("getting tag %w", err)
		}
		if tag == nil {
			if _, err = s.TagService.CreateTag(tagName, nil); err != nil {
				return fmt.Errorf("creating tag %w", err)
			}
		}
	}
	return nil
}
